### PR TITLE
Created CI Environment Base

### PR DESCRIPTION
**Pull request by fsw0422**
Currently only unit test is running but in the future, it can incorporate static analysis, integration test etc.

The main focus of the environment setup are
  - Maximize portability across all platforms (Windows / OSX / Linux), only using Python3
  - CI technology agnostic (treating Github Actions as a simple entrypoint, steps graph and machine resource). This means each steps in CI runs in the exact same container environment, either locally or in Github Actions

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+name: ci 
+on: push
```
Comment (fsw0422) [inline .github/workflows/ci.yml:2]: This may need to get removed since we may run out of quota in the future if we merge alot of changes
https://docs.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions#about-billing-for-github-actions

**Comment by gak** [comment]
I have a high aversion to using python as a dependency (but also any other dependencies). It has a horrible developer experience.

I'm sorry but unless there's a good reason to use python, I have to deny this PR.

**Comment by gak** [comment]
Please in the future let me know your plans so we don't waste PRs in the future.

**Comment by fsw0422** [comment]
I understand the maintenance nightmare, but personally I went through nightmare only when maintaining more than thousands of lines of code (Would never use Python for web services again).

Using Rust for the CI, will complicate the CI env develop iteration since we need another step for building the CI code everytime it changes. Correct me if I'm wrong but Rust also doesn't provide sophisticated build tool specific language feature like Gradle or SBT where you can incorporate your CI process in the build file.

I'm also considering to write the publish code for semver auto incrementation much like the Maven release plugin. It requires tagging, version incremental auto committing, packaging, containerizing and finally publishing in a single command "atomically". So far to my knowledge, I don't know if there's a better language than Python for this specific task

**Comment by gak** [comment]
I'm looking into this and trying to work out what the goal here is?

In my mind I just want to run github actions with the commands in that Makefile, and as you mentioned in the future automatically increment the version, tag it, and generate+push a docker image. These can all be done with existing github actions and cargo tools if I'm not mistaken.

I've set up a github action to run the Makefile which is running now, which seems to do what this PR currently does.

**Comment by fsw0422** [comment]
the build itself doesn't need to run in all environments since the packaged binary just needs to be published in an alpine Linux image (the goal of containerization). What I meant by cross-platform was being independent of UNIX tools (Bash, Makefile etc) that's why the PR contains Python to make the entrypoint itself to be cross-platform. In that sense, adding any other Python package than docker does not make sense and it will stay that way forever, plus the PR code will also hardly ever change since it's just an entrypoint to ignite the entire CICD process

another goal was to make it as technology agnostic as possible. I see that in the current github workflow, we're relying on quite of features provided by Github Actions. In the past, I've had trouble migrating CI from CI for the sake of just trying out new technology or due to price change and quota from the provider. That's why I've built buildenv in the PR. this image encapsulates all dependencies needed for building and publishing feeless and it's platform independent since it's using only Docker. It simulates the exact same behavior regardless of which machine you run on (means the CI is running the exact same environment and code as local environment by the entrypoint command, which is test.py in PR).  

For release, we can use the same buildenv image and use multistage Docker command to package / publish it

Tldr, I'm trying to treat the CI environment no more than the stage graph resolver and machine resource (probably change detector to trigger the pipeline). If we can encapsulate everything other than that into the project code, it's separate the concern well. 

but on the bottom-line, it's my opinion and it's your decision as you lead the project. I'm happy to go with any approach you take as well

**Comment by gak** [comment]
I see what your goal is now and understand. I think you have a good plan to make the platform not reliant on CI, but for now let's just live on GHA. If we need to change to another CI system we can continue your buildenv effort.

Thanks for your efforts by the way, I really don't like seeing great work being wasted. We'll leave this PR around for the future.
